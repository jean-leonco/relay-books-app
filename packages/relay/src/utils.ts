import AsyncStorage from '@react-native-community/async-storage';

import { AUTH_KEY, I18N_KEY } from './config';

export const getToken = async () => {
  return AsyncStorage.getItem(AUTH_KEY);
};

export const clearToken = async () => {
  return AsyncStorage.removeItem(AUTH_KEY);
};

export const getLanguage = async () => {
  return AsyncStorage.getItem(I18N_KEY);
};
