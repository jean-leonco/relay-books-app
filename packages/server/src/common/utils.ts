import { Model, Types } from 'mongoose';
import { fromGlobalId } from 'graphql-relay';

// https://stackoverflow.com/a/2593661/710693
export const escapeRegex = (str: string) => `${str}`.replace(/[.?*+^$[\]\\(){}|-]/g, '\\$&');

export const PLATFORM = {
  APP: 'APP',
  UNKNOWN: 'UNKNOWN',
};

export const getPlatform = (platform?: string | null): string => {
  if (typeof platform !== 'string') {
    return PLATFORM.UNKNOWN;
  }

  //if (platform.includes('ios') || platform.includes('android')) {
  //  return PLATFORM.APP;
  //}

  if (platform === PLATFORM.APP) {
    return PLATFORM.APP;
  }

  return PLATFORM.UNKNOWN;
};

export function kilometersToEarthRadians(kilometers: number) {
  return kilometers / 6378.1;
}

// returns an ObjectId given an param of unknown type
export const getObjectId = (target: string | Model<any> | Types.ObjectId | undefined | null): Types.ObjectId | null => {
  if (!target) {
    return null;
  }

  if (target instanceof Types.ObjectId) {
    return new Types.ObjectId(target.toString());
  }

  if (typeof target === 'object') {
    return target && target._id ? new Types.ObjectId(target._id) : null;
  }

  if (typeof target === 'string') {
    const result = fromGlobalId(target);

    if (result.type && result.id && Types.ObjectId.isValid(result.id)) {
      return new Types.ObjectId(result.id);
    }

    if (Types.ObjectId.isValid(target)) {
      return new Types.ObjectId(target);
    }

    return null;
  }

  return null;
};

type EnumObject = {
  [key: string]: string;
};

type EnumObjectResult = {
  [key: string]: {
    value: string;
  };
};

export const enumBuilderValues = (constants: EnumObject): EnumObjectResult =>
  Object.keys(constants).reduce(
    (prev, curr) => ({
      ...prev,
      [curr]: {
        value: constants[curr],
      },
    }),
    {},
  );
