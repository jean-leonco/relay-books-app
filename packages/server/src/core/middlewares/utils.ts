import { GraphQLResolveInfo } from 'graphql';
import { ObjectSchema, Shape } from 'yup';

import { GraphQLContext } from '../../types';

interface GraphQLMutationConfig {
  validationSchema?(context: GraphQLContext): ObjectSchema<Shape<any, any>>;
  authenticatedOnly?: boolean;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): GraphQLMutationConfig | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
