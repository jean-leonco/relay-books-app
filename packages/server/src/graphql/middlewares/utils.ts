import { GraphQLField, GraphQLResolveInfo } from 'graphql';

import { Extensions } from '../../types';

interface Field extends GraphQLField<any, any> {
  extensions: Extensions | null | undefined;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): Field | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
