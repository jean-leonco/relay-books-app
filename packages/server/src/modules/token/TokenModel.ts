import mongoose, { Document, Model } from 'mongoose';

import { IStatusSchema, ObjectId } from '../../types';

import { PLATFORM } from '../../security';
import isActiveMongooseField from '../../mongoose/isActiveMongooseField';

export enum TOKEN_SCOPES {
  RESET_PASSWORD = 'BOOKSAPP:RESET_PASSWORD', // used when user requested a password reset
  SESSION = 'BOOKSAPP:SESSION', // token for a valid user session
}

const Schema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      description: 'The user owner of this token.',
      index: true,
      default: null,
    },
    ip: {
      type: String,
      description: 'Agent IP that came from request headers.',
      required: true,
      index: true,
    },
    platform: {
      type: String,
      required: true,
      default: PLATFORM.APP,
      description: 'Channel that this token belongs to. ex: APP.',
      enum: Object.values(PLATFORM),
    },
    scope: {
      type: String,
      description: 'Scope of this token. ex: SESSION.',
      required: true,
      index: true,
      enum: Object.values(TOKEN_SCOPES),
    },
    expiresIn: {
      type: Date,
      description: 'When this token expires.',
      default: null,
    },
    isBlocked: {
      type: Boolean,
      description: 'If the token is blocked.',
      default: false,
      index: true,
    },
    ...isActiveMongooseField,
  },
  {
    timestamps: true,
    collection: 'Token',
  },
);

export interface IToken extends Document, IStatusSchema {
  userId?: ObjectId;
  ip: string;
  platform: PLATFORM;
  scope: TOKEN_SCOPES;
  expiresIn: Date | null;
  isBlocked: boolean;
}

const TokenModel: Model<IToken> = mongoose.model('Token', Schema);

export default TokenModel;
