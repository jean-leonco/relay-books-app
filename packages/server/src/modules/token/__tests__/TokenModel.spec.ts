import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  sanitizeTestObject,
} from '@workspace/test-utils';

import { createToken, createUser } from '../../../test/utils';

import TokenModel, { IToken, TOKEN_SCOPES } from '../TokenModel';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('TokenModel', () => {
  it('should be able to create a new token using TokenModel', async () => {
    const user = await createUser();
    const ip = '::ffff:127.0.0.1';
    const expiresIn = new Date('11/23/2020');
    const scope = TOKEN_SCOPES.SESSION;

    const token = await new TokenModel({
      userId: user._id,
      ip,
      scope,
      expiresIn,
    }).save();

    const tokenObj = await TokenModel.findOne({ _id: token._id }).lean<IToken>();

    expect(tokenObj?.userId).toMatchObject(user._id);
    expect(tokenObj?.ip).toBe(ip);
    expect(tokenObj?.scope).toBe(scope);
    expect(tokenObj?.expiresIn).toMatchObject(expiresIn);
    expect(tokenObj?.isBlocked).toBe(false);
    expect(sanitizeTestObject(tokenObj)).toMatchSnapshot();
  });

  it('should be able to create a new token without user', async () => {
    const ip = '::ffff:127.0.0.1';
    const expiresIn = new Date('11/23/2020');
    const scope = TOKEN_SCOPES.SESSION;

    const token = await new TokenModel({
      ip,
      scope,
      expiresIn,
    }).save();

    const tokenObj = await TokenModel.findOne({ _id: token._id }).lean<IToken>();

    expect(tokenObj?.userId).toBe(null);
    expect(tokenObj?.ip).toBe(ip);
    expect(tokenObj?.scope).toBe(scope);
    expect(tokenObj?.expiresIn).toMatchObject(expiresIn);
    expect(tokenObj?.isBlocked).toBe(false);
  });

  it('should be able to create a new token without expiresIn', async () => {
    const user = await createUser();
    const ip = '::ffff:127.0.0.1';
    const scope = TOKEN_SCOPES.SESSION;

    const token = await new TokenModel({
      userId: user._id,
      ip,
      scope,
    }).save();

    const tokenObj = await TokenModel.findOne({ _id: token._id }).lean<IToken>();

    expect(tokenObj?.userId).toMatchObject(user._id);
    expect(tokenObj?.ip).toBe(ip);
    expect(tokenObj?.scope).toBe(scope);
    expect(tokenObj?.expiresIn).toBe(null);
    expect(tokenObj?.isBlocked).toBe(false);
  });

  it('should be able to create a new token using createRow', async () => {
    const ip = '::ffff:127.0.0.1';
    const scope = TOKEN_SCOPES.SESSION;

    const token = await createToken();

    const tokenObj = await TokenModel.findOne({ _id: token._id }).lean<IToken>();

    expect(tokenObj?.ip).toBe(ip);
    expect(tokenObj?.scope).toBe(scope);
    expect(tokenObj?.expiresIn).toBe(null);
    expect(tokenObj?.isBlocked).toBe(false);
  });

  it('should be able to create a new token with custom properties', async () => {
    const ip = '192.168.0.117';
    const scope = TOKEN_SCOPES.RESET_PASSWORD;

    const token = await createToken({ ip, scope });

    const tokenObj = await TokenModel.findOne({ _id: token._id }).lean<IToken>();

    expect(tokenObj?.ip).toBe(ip);
    expect(tokenObj?.scope).toBe(scope);
    expect(tokenObj?.expiresIn).toBe(null);
    expect(tokenObj?.isBlocked).toBe(false);
  });
});
