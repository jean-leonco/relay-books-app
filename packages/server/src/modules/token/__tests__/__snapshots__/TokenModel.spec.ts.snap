// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenModel should be able to create a new token using TokenModel 1`] = `
Object {
  "__v": 0,
  "_id": "ObjectId",
  "createdAt": "FROZEN-CREATEDAT",
  "expiresIn": 2020-11-23T00:00:00.000Z,
  "ip": "::ffff:127.0.0.1",
  "isActive": true,
  "isBlocked": false,
  "platform": "APP",
  "scope": "BOOKSAPP:SESSION",
  "updatedAt": "FROZEN-UPDATEDAT",
  "userId": "ObjectId",
}
`;
